src/analysis/syntax_doc.ml

```diff
@@ -0,0 +1,280 @@
+            documentation =
+              "https://v2.ocaml.org/releases/5.1/htmlman/privatetypes.html#ss:private-types-variant";
+          }
+    | (_, Type_kind Ttype_open)
+      :: (_, Type_declaration { typ_private = Private; _ })
+      :: _ ->
+        Some
+          {
+            name = "Private Extensible Types";
+            description =
+              "Enable libraries to reveal some, but not all aspects of the \
+               implementation of a type to clients of the library";
+            documentation =
+              "https://v2.ocaml.org/releases/5.1/htmlman/privatetypes.html#start-section";
+          }
+    | (_, Type_kind Ttype_abstract)
+      :: (_, Type_declaration { typ_private = Private; _ })
+      :: _ ->
+        Some
+          {
+            name = "Private Type Abbreviations";
+            description =
+              "A private type abbreviation declares a type that is distinct \
+               from its implementation type `typexpr`.";
+            documentation =
+              "https://v2.ocaml.org/releases/5.1/htmlman/privatetypes.html#ss:private-types-abbrev";
+          }
+    | (_, Expression _)
+      :: (_, Expression _)
+      :: (_, Value_binding _)
+      :: (_, Structure_item ({ str_desc = Tstr_value (Recursive, _); _ }, _))
+      :: _ ->
+        Some
+          {
+            name = "Recursive definitions of values";
+            description =
+              "The `let rec` binding construct, in addition to the definition \
+               of recursive functions, also supports a certain class of \
+               recursive definitions of non-functional values, such as `let \
+               rec name1 = 1 :: name2 and name2 = 2 :: name1 in expr` which \
+               binds `name1` to the cyclic list `1::2::1::2::…`, and `name2` \
+               to the cyclic list `2::1::2::1::…`";
+            documentation =
+              "https://v2.ocaml.org/releases/5.1/htmlman/letrecvalues.html";
+          }
+    | [ (_, Structure _) ] ->
+        Some
+          {
+            name = "Documentation comments";
+            description =
+              "Automatically converted during parsing into attributes to allow \
+               tools to process them as documentation";
+            documentation =
+              "https://v2.ocaml.org/releases/5.1/htmlman/doccomments.html";
+          }
+    | (_, Module_expr _)
+      :: (_, Module_type { mty_desc = Tmty_typeof _; _ })
+      :: _ ->
+        Some
+          {
+            name = "Recovering the type of a module";
+            description =
+              "The construction `module type of module-expr` expands to the \
+               module type (signature or functor type) inferred for the module \
+               expression `module-expr`. ";
+            documentation =
+              "https://v2.ocaml.org/releases/5.1/htmlman/moduletypeof.html";
+          }
+    | node -> (
+        let n = List.rev node in
```

Comment:
I was intrigued by this `List.rev` and the matching that follows. It has two consequences:

- It always find the recursive module for every cursor position inside it. But we only want to show the documentation when the cursor is on the `rec` keyword (and maybe the `module` keyword)

- It won't match on nested recursive modules: 
```
module B = struct 
  module rec A : sig 
  ...
  end
end
```

I don't know what merlin identifies as the "node under the cursor" when on the `rec` keyword, but if this is not what we need then we should change that :-)